indexers: asynchronous index updates.

dnldd opened the pull request:
This reworks index updates by introducing an index subscriber which allows optional indexers to be asynchronously updated. The Index manager which synchronously updated optional indexes has been removed.

Resolves #1470.

dnldd commented:
I didn't factor in recover on startup via the tip height, that's why I'm tracking the tip height here and not the hash, will correct. My plan is to update the other optional indexes to track the tip and update the index manager to asynchronously receive updates for processing by enabled indexes.

davecgh commented:
I'll wait a bit until you get further before making a final judgement, but, my gut feeling is that the approach of trying to make the manager asynchronous with a message channel for connect/disconnect is not going to work out super well.  Without going all of the various reasons why, I think it's sufficient to consider that the primary issue is likely to be that it is still relying on being able to send connect/disconnect messages from within the database transaction and, even though it does the indexing of that particular block asynchronous, it will just end up blocking on the channel send when the next block is processed if the previous index operation is ongoing.

More generally, the existing infrastructure is designed specifically for synchronous indexing and under the assumption that it all happens under an atomic database transaction.  The entire goal #1470 is to make that no longer required, which means just making the existing infrastructure run in a goroutine isn't likely to work well, since it will almost positively lead to needing a bunch of less than ideal hacks and can kicking to make it work instead of actually having a solid asynchronous design.

Further, there are indexes which are required to be in sync for proper consensus operation, such as cfilterv2, so you can't really just make them all async.

dnldd commented:
Noted, thanks for taking a look. Will rethink my approach and tinker some more.

dnldd commented:
Coverages for updated rpc tests.

```sh
❯ go tool cover -func=cov.out | grep -E "handleExistsAddress"
github.com/decred/dcrd/internal/rpcserver/rpcserver.go:1479:    handleExistsAddress              100.0%


❯ go tool cover -func=cov.out | grep -E "handleExistsAddresses" 
github.com/decred/dcrd/internal/rpcserver/rpcserver.go:1531:    handleExistsAddresses            100.0%


❯ go tool cover -func=cov.out | grep -E "handleGetCFilter" 
github.com/decred/dcrd/internal/rpcserver/rpcserver.go:2318:    handleGetCFilter         100.0%

❯ go tool cover -func=cov.out | grep -E "handleGetCFilterHeader"github.com/decred/dcrd/internal/rpcserver/rpcserver.go:2387:    handleGetCFilterHeader           100.0%

❯ go tool cover -func=cov.out | grep -E "handleSearchRawTransactions"
github.com/decred/dcrd/internal/rpcserver/rpcserver.go:4382:    handleSearchRawTransactions      99.3%
```

dnldd commented:
`handleGetRawTransaction` does not have an rpc test yet, will get that done and make the needed modifications.

rstaudt2 commented:
I haven't really gotten into the review yet, but I kicked off some testing with these changes, and `dcrd` seems to hang for me.  When kicking off a sync on mainnet, it gets stuck right when starting to sync blocks.  This happens for me whether indexes are enabled or not.  When shutting down `dcrd` at this point, I also noticed it hangs trying to shut down the address manager, and I have to force quit.
```
...
2021-02-26 04:43:02.814 [INF] SYNC: Initial headers sync complete (best header hash 000000000000000008a96a69c680d2481500c1a3c543f2722d2d7667bbecd8f2, height 532167)
2021-02-26 04:43:02.814 [INF] SYNC: Syncing chain
(stuck here)
```

rstaudt2 commented:
Some of this will probably be obvious after reviewing most of the code, which I'm just getting started on, but given this is a pretty large set of changes, I think it would be helpful both for current reviewers and for historical purposes to provide a more detailed description of the changes in the PR description, commit messages, and/or comments in the code.

Specifically, I think it would be good to describe:
- The high-level design for the changes, including the concurrency pattern being used
- The performance impact of the changes
- Summary of the new data structures being introduced and how they are used (`IndexSubscription`, `IndexSubscriber`, `prevScriptsSnapshot`, etc.)
- etc.

rstaudt2 commented:
I haven't tested this yet due to the issue I was running into above, but one thing that I wanted to point out is that since an unbuffered channel is being used to notify the indexes when connecting/disconnecting blocks, it is going to end up blocking when sending on that channel if the indexer is busy processing a prior notification (and therefore prevent any further opportunity to run in parallel). 

This could be alleviated by using a buffered channel but would need to determine an appropriate capacity and look at the overall performance in practice.  It _should_ improve the block connection performance and reduce the overall sync time since it would have the opportunity to run in parallel (if multiple CPUs are available) as well as context switch on blocking I/O, but those are things we'll want to evaluate in practice.

rstaudt2 commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +171,788 @@ func interruptRequested(ctx context.Context) bool {
+	case s.c <- ntfn:
+	default:
+		// Non-blocking send fallthrough.
+	}
```
This causes notifications to be missed by the subscribers when the buffer is full, and therefore the index will be missing data.  I hit this case even with the buffer size of 128, but you can more easily simulate this by changing the buffer size to 1.  If you then attempt to query for an address/tx within that missed block, you'll see that the data is missing, even though the data may be there for later blocks.

The current assumption/design is the blocks are connected to the indexers in order (since, for example, if the tip height is 100, the indexer is assuming that all blocks up to 100 were connected).  In that case, you may want to error if a block is connected and the height is not the current tip height plus one so that this wouldn't be silently missed.

Unless you go with an alternative design, you'll need to block when the buffer is full here since the indexers currently rely on the notifications for each block.

rstaudt2 reviewed (changes requested):
I can confirm that the issue I saw above (where syncing would hang) is now resolved.

One issue I noticed in the new version (commented further inline) is that notifications are being missed resulting in missing data in the indexes.  I still have a lot more to review/test but left a few initial comments inline.

rstaudt2 reviewed (changes requested):
Thanks for the updates so far.  I was able to fully sync with indexes enabled now, and the indexes seem to be working as expected.  I'll be kicking off more performance and edge case testing as I continue the review.

dnldd commented:
Nice, will also pivot to getting the benchmarks done. Thinking of using a harness to generate blocks and sync to SVH for both setups and comparing the results.

dnldd commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +171,788 @@ func interruptRequested(ctx context.Context) bool {
+	// Only relay notifications when there are subscribed indexes
+	// to be notified.
+	if clients > 0 {
+		s.c <- ntfn
```
I see what you're referring to however the `IndexSubscriber` would need a goroutine to coordinate the shutdown process for the indexes instead of passing the server context individually to each index. As a result relaying index notifications is also being done there. Relaying index notifications from `Notify` could the block the block connection/disconnection process for every index being notified as well.

rstaudt2 replied:
OK, no major issue, just thought that it's maybe a little harder to conceptually follow.

It is still going to ultimately block on `Notify`, it's just that you have two layers of buffers.  For example:

- If an `IndexSubscription` buffer is full, then the `IndexSubscriber` is going to block in `publishIndexNtfn` (where it writes to that `IndexSubscription` channel).
- If `IndexSubscriber` is blocked, it's not going to be dequeuing from its own buffer, which will then fill up.
- When the `IndexSubscriber` buffer is full, `Notify` will block when trying to write to the channel.

This isn't an issue since it has to block at some point if the index gets far enough behind, but just clarifying that point,

dnldd commented on internal/rpcserver/rpcserver.go:
```diff
@@ -1517,7 +1521,31 @@ func handleExistsAddress(_ context.Context, s *Server, cmd interface{}) (interfa
+			return nil, rpcInternalError(msg, "Sync")
+		case <-existsAddrIndex.WaitForSync():
+			break sync
+		}
```
Blocking in the RPC handler ensures the response i guaranteed to be accurate to the node's current chain state and isn't being derived from stale data. We are also blocking for a maximum of `syncWait` here (3 seconds max currently). If the sync takes too long an error will be returned.

Here is a snippet from `handleGetRawTransaction`:

```go
sync:
for !chain.BestSnapshot().Hash.IsEqual(tHash) {
	select {
	case <-time.After(syncWait):
		msg := fmt.Sprintf("%s: index not synced", txIndex.Name())
		return nil, rpcInternalError(msg, "Sync")
	case <-txIndex.WaitForSync():
		break sync
	}
}

...
```

davecgh replied:
I'm a bit torn on this.  I generally agree with @rstaudt2 that blocking in RPC handlers is typically not super desirable because it results in a laggy UX and can even break some speed guarantees that things such as the [Decred Rosetta middleware, dcrros](https://github.com/decred/dcrros) have to make.

On the other hand, callers, including the aforementioned `dcrros`, currently have the expectation that they can make queries based on the current state of the chain once they receive a `blockconnected` notification.  Without some logic like this, all of those callers who have relied on that implicit assumptions for 5+ years will suddenly break.

Along those same lines, something else that needs to be checked here is that there isn't any **internal** code in dcrd that relies on such behavior.  I don't believe there is, but it's possible something in the fee estimator code might.

In any case, given the above, I think the main two choices are to either:

a) Keep this, admittedly rather hacky, logic to ensure continuity; or
b) Separate blockchain notifications from the ultimate RPC notifications where the latter are not sent until continuity is achieved.

I think that ultimately (b) might be a better option, but it will be tricky to implement properly and probably should be separate from this PR.  For example, such logic would need to deal with cases where there are no indexers active and thus can send the notification immediately and cases where maybe some of the indexers are caught up and others aren't, among others.

One benefit of the approach in this PR that is kind of nice which option (b) wouldn't provide is that it won't block RPCs that don't care about certain indexes even if those indexes are not yet current.  For example, `getrawtransaction` won't block if the `txindex` is already synced and the `addrindex` is still working.

rstaudt2 replied:
> On the other hand, callers, including the aforementioned dcrros, currently have the expectation that they can make queries based on the current state of the chain once they receive a blockconnected notification. Without some logic like this, all of those callers who have relied on that implicit assumptions for 5+ years will suddenly break.

The current logic will break this assumption if the delay to asynchronously update the indexes is greater than the configured wait time (currently 3 seconds).  It's probably not likely to hit this case since we are limiting the channel buffer size, but something to consider depending on how far ahead of the indexers the chain can get during IBD.

Outside of that, I agree with your assessment and that it's probably best to keep it as is.

dnldd commented on blockchain/indexers/common.go:
```diff
@@ -27,6 +31,16 @@ var (
+	noPrereqs = "none"
+
+	// bufferSize represents the index notification buffer size.
+	bufferSize = 128
```
I just picked `128` there, I do realize on initial sync this gets filled very quickly considering the stream of blocks coming in however providing a buffer that caters to the worst case scenario here (initial node sync, thousands of blocks streaming in) would be overkill for almost all other scenarios however (synced node receiving a new block, a node a few days out-of-sync, a node a few weeks out of sync). Perhaps the buffer could be resized once the node is synced? I'm open to suggestions.

dnldd commented on blockchain/chain.go:
```diff
@@ -2038,53 +2126,6 @@ func stxosToScriptSource(block *dcrutil.Block, stxos []spentTxOut, isTreasuryEna
-// associated versions spent by the given block by using the spend journal.
-//
-// It is defined via a separate internal struct to avoid polluting the public
-// API of the BlockChain type itself.
```
I didn't think it was worth creating a new adapter considering `Blockchain` implements a majority of the functions required by the `ChainQueryer` interface. And I wouldn't say the added functions `Best` and `PrevScripts` are polluting the public API of `Blockchain`, disregarding the `ChainQueryer` requirement they do belong in the public API of `Blockchain` I think.

rstaudt2 replied:
We can let others weigh-in, I mostly called it out due to the existing comments in the code.  One thing to note is that `PrevScripts` returns an indexing specific interface (`indexers.PrevScripter`), and if this is part of the public BlockChain API, it may be used for something unrelated to indexing.

davecgh replied:
I commented on this above before seeing it was removed, but I agree with @rstaudt2.  I really don't want to see the public API polluted with things that are not specific to the blockchain.

davecgh replied:
In fact, I ultimately would like to `blockchain` made internal, but there are some things that external callers still rely on which would to be make available via some other means.

dnldd commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +172,776 @@ func interruptRequested(ctx context.Context) bool {
+	// associated with this subscription to have processed incoming
+	// notifications before it does. A nil dependency indicates the subscription
+	// has no dependencies.
+	dependency *IndexSubscription
```
Noted, updating.

rstaudt2 replied:
I think the last part `A nil dependency indicates the subscription has no dependencies` needs to be updated as well.  I think something like `If dependent is nil, then the subscription has no dependent subscription` would be more accurate.

dnldd commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +172,776 @@ func interruptRequested(ctx context.Context) bool {
+
+	// If the subscription has a dependency, stop it as well.
+	if s.dependency != nil {
+		err := s.dependency.Stop()
```
Noted, will unexport the function. Thanks. Regarding the mutex, `Stop` currently has a single call site (at `handleNotifications` when signalling the indexes to shutdown. The mutex is held before looping through the subscriptions, since the the dependent's `Stop` is called in a recursively the initial mutex held still applies.

rstaudt2 replied:
I see.  I think it was the fact that it requires holding the `subscriber` mutex, and I read it as the `subscription` mutex.  That looks fine then.

This is more of a stylistic preference, but I think it would make more sense for this function to live on the `IndexSubscriber`.  Since the `IndexSubscriber` owns the `subscriptions` map, I think it should be the only one modifying it.

rstaudt2 commented on blockchain/indexers/manager.go:
```diff
@@ -1,874 +0,0 @@
-	// tip and need to be caught up, so log the details and loop through
-	// each block that needs to be indexed.
-	log.Infof("Catching up indexes from height %d to %d", lowestHeight,
-		bestHeight)
```
This is still missing the logs that I was referring to.  For example, on `master` it will show up as follows:
```
2021-03-05 05:40:46.659 [INF] INDX: Catching up indexes from height 318967 to 445659
2021-03-05 05:40:56.660 [INF] INDX: Indexed 6899 blocks in the last 10s (117111 transactions, height 325866, 2019-03-11 20:12:20 -0500 CDT)
2021-03-05 05:41:06.660 [INF] INDX: Indexed 7528 blocks in the last 10s (123832 transactions, height 333394, 2019-04-06 16:17:00 -0500 CDT)
...
```

rstaudt2 commented on blockchain/indexers/addrindex_test.go:
```diff
@@ -274,3 +283,448 @@ nextTest:
+		IsTreasuryEnabled: false,
+	}
+	subber.Notify(ntfn)
+	time.Sleep(time.Millisecond * 150)
```
Is this sleep intending to wait long enough for the notification to be processed?  That would likely lead to flakey tests that intermittently fail when it takes longer than that amount of time.

dnldd commented:
```
 -- sync for v1.6.0 --
❯ dcrd --txindex    
2021-03-05 14:43:53.517 [INF] DCRD: Version 1.6.0+release.local (Go version go1.15.6 linux/amd64)
2021-03-05 14:43:53.517 [INF] DCRD: Home dir: /home/dnldd/.dcrd
2021-03-05 14:43:53.517 [INF] DCRD: Loading block database from '/home/dnldd/.dcrd/data/mainnet/blocks_ffldb'
2021-03-05 14:43:53.544 [INF] DCRD: Block database loaded
2021-03-05 14:43:53.545 [INF] INDX: Transaction index is enabled
2021-03-05 14:43:53.545 [INF] INDX: Exists address index is enabled
2021-03-05 14:43:53.545 [INF] INDX: CF index is enabled
2021-03-05 14:43:53.565 [INF] CHAN: Loading block index...
2021-03-05 14:43:57.790 [INF] STKE: Stake database version 1 loaded
2021-03-05 14:43:57.793 [INF] INDX: Catching up indexes from height 0 to 534247
2021-03-05 14:44:07.795 [INF] INDX: Indexed 27523 blocks in the last 10s (386002 transactions, height 27523, 2016-05-13 00:20:33 +0000 GMT)
2021-03-05 14:44:17.795 [INF] INDX: Indexed 20844 blocks in the last 10s (355649 transactions, height 48367, 2016-07-24 10:40:10 +0000 GMT)
2021-03-05 14:44:27.795 [INF] INDX: Indexed 16992 blocks in the last 10s (285259 transactions, height 65359, 2016-09-21 15:14:36 +0000 GMT)
...
2021-03-05 14:57:13.138 [INF] INDX: Indexed 2838 blocks in the last 10.1s (53020 transactions, height 520968, 2021-01-18 12:16:42 +0000 GMT)
2021-03-05 14:57:23.138 [INF] INDX: Indexed 9040 blocks in the last 10s (171731 transactions, height 530008, 2021-02-18 21:43:58 +0000 GMT)
2021-03-05 14:57:33.102 [INF] INDX: Indexes caught up to height 534247

duration: approx: 13 minutes.


--- sync for async-index ---
~/Work/dcrd track_tip_txindex* ⇣⇡
❯ ./async-index --txindex
2021-03-07 11:42:11.892 [INF] DCRD: Version 1.7.0-pre (Go version go1.15.6 linux/amd64)
2021-03-07 11:42:11.895 [INF] DCRD: Home dir: /home/dnldd/.dcrd
2021-03-07 11:42:11.896 [INF] DCRD: Loading block database from '/home/dnldd/.dcrd/data/mainnet/blocks_ffldb'
2021-03-07 11:42:11.955 [INF] DCRD: Block database loaded
2021-03-07 11:42:11.997 [INF] CHAN: Loading block index...
2021-03-07 11:42:16.339 [INF] STKE: Stake database version 1 loaded
2021-03-07 11:42:16.347 [INF] CHAN: UTXO cache initializing (max size: 150 MiB)...
2021-03-07 11:42:16.349 [INF] CHAN: UTXO cache initialization completed
2021-03-07 11:42:16.349 [INF] CHAN: Blockchain database version info: chain: 9, compression: 1, block index: 3
2021-03-07 11:42:16.349 [INF] CHAN: Best known header: height 534249, hash 00000000000000000d2401acd13fde402c370e15e22b630926abb76138d8a46d
2021-03-07 11:42:16.349 [INF] CHAN: Chain state: height 534249, hash 00000000000000000d2401acd13fde402c370e15e22b630926abb76138d8a46d, total transactions 9210999, work 29354304413087377840917024, progress 100.00%
2021-03-07 11:42:16.349 [INF] INDX: Transaction index is enabled
2021-03-07 11:42:16.349 [INF] INDX: transaction index: catching up from from height 0 to 534249
2021-03-07 11:42:26.350 [INF] INDX: Indexed 21165 blocks in the last 10s (277231 transactions, height 21165, 2016-04-21 03:00:50 +0000 GMT)
2021-03-07 11:42:36.350 [INF] INDX: Indexed 17437 blocks in the last 10s (298927 transactions, height 38602, 2016-06-20 13:08:23 +0000 GMT)
....
2021-03-07 11:54:00.525 [INF] INDX: Indexed 6837 blocks in the last 10s (122660 transactions, height 525399, 2021-02-02 23:11:30 +0000 GMT)
2021-03-07 11:54:10.552 [INF] INDX: Indexed 7646 blocks in the last 10.02s (163672 transactions, height 533045, 2021-03-01 13:06:40 +0000 GMT)
2021-03-07 11:54:15.911 [INF] INDX: transaction index: index caught up to height 534249

duration: approx 13 minutes.
```

The current design is less efficient particularly with regards to catching up the indexes because the process is being run by each index whereas the `IndexManager` handled it for all indexes it managed. I plan on making some modifications to get as close to that as possible.

For recovering indexes, I plan on having the indexes do not have any prerequisites check if their current tips are on the main chain, if not they will process disconnection notification and relay them to their dependent indexes until their tips are back on the main chain.

For catching up indexes, I plan on having having the `IndexSubscriber` (will most likely rename it back to `IndexManager` since its doing more  now) handle it by getting all the index tips to determine the lowest height synced to among indexes and relaying updates from that point to the current chain tip.

rstaudt2 commented on blockchain/indexers/common.go:
```diff
@@ -27,6 +32,16 @@ var (
+	noPrereqs = "none"
+
+	// bufferSize represents the index notification buffer size.
+	bufferSize = 1
```
It looks like you may have been doing some testing with a buffer size of 1 and forgot to change it back.

dnldd replied:
Ahh missed it again, resolving.

rstaudt2 commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +182,900 @@ func interruptRequested(ctx context.Context) bool {
+	if expectedHeight != ntfn.Block.Height() {
+		log.Tracef("%s: relaying notification for height %d to dependent",
+			indexer.Name(), ntfn.Block.Height())
+		return notifyDependent(ctx, indexer, ntfn)
```
If the notification isn't for the expected height, why does it just notify the dependent and move on?  Wouldn't this result in a missed notification for this index?

dnldd replied:
For index notifications that are less than the index tip height its fine to just relay the notification to the dependent since the dependent could be synced to a lower tip height. I do agree the check needs refinement, it needs to error if the index notification received is greater than the expected height.

rstaudt2 commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +182,900 @@ func interruptRequested(ctx context.Context) bool {
+		if !queryer.MainChainHasBlock(tipHash) {
+			return fmt.Errorf("%s: index tip (%s) is not on the main chain",
+				sub.idx.Name(), tipHash)
+		}
```
I don't think we can error out in this case.  The index tip could be on a side chain block, and it still needs to recover properly.  I had to handle this case in the [UtxoCache initialization code](https://github.com/decred/dcrd/blob/8d4ac771383f254b7a68cd89aee37e2ac19de18e/blockchain/utxocache.go#L707-L880) as well, which requires disconnecting blocks back to the fork point before replaying the main chain blocks to catch up.

dnldd replied:
The indexes currently recover to the main chain on initialization if needed via `recover`. If `Catchup` encounters an index tip not on the main chain that'd mean the recovery process failed and should error as a result.

rstaudt2 replied:
Got it, thanks.  I missed that that was handled separately.

rstaudt2 commented on internal/rpcserver/rpcserver.go:
```diff
@@ -2748,16 +2800,40 @@ func handleGetRawTransaction(_ context.Context, s *Server, cmd interface{}) (int
+		}
+
+		chain := s.cfg.Chain
+		if chain.BestSnapshot().Height > (tHeight + 5) {
```
We may want to add a comment here and/or define `5` as a constant with a description.

rstaudt2 commented on blockchain/indexers/common.go:
```diff
@@ -173,3 +182,909 @@ func interruptRequested(ctx context.Context) bool {
+			// of the block if the index requires them.
+			if indexNeedsInputs(idx) {
+				var err error
+				prevScripts, err = queryer.PrevScripts(dbTx, block)
```
There is one issue with the recovery code.  This panics when attempting to fetch the spend journal because it will no longer exist since the spend journal is removed when disconnecting blocks.

You can simulate this easily by:

- Enabling indexes and start syncing
- Restart `dcrd` with indexes disabled
- Use `invalidateblock` to invalidate a block at a previous height
- Restart `dcrd` with indexes enabled
- The index will need to rollback to the block before the invalidated block (where the tip currently is).  It will crash since the spend journal is not available for those blocks that it needs to roll back.

dnldd replied:
After going back and forth on how best to resolve this issue I think the best way forward would be to track spend journal entry removals and only remove them once they are out of range for a reorg per the max reorg depth. This would also mean blocks connected to the chain will have to ensure their associated spend journals are not scheduled to be removed, if they are the action is cancelled since the block has been reconnected back to the chain. This tracking would need to be persisted to db. If there are no objections or suggestions of a better solution I'll proceed to implement this in a different PR.

rstaudt2 replied:
Can you summarize the solution in the current PR around this?  This comment made it sound like you wanted to defer fixing this, but I see a bunch of logic for a `SpendJournalFilterer` was added.  So I just want to make sure I understand the approach before I continue reviewing.

dnldd replied:
Sure, the solution was to defer the removal of the spend journal entries for the blocks being disconnected since its possible they will be needed later in a recovery process. For block invalidation, all block hashes being disconnected are added to the prune filter to prevent their removal. The prune filter is updated on startup particularly for indexes that need inputs from transactions. The filter keeps count of the number of processes that require the spend journal associated with the block hash by incrementing the count when `AddEntry` is called, `RemoveEntry` decrements the count until there are no processes needing it, at that point the spend journal entry is finally deleted.

rstaudt2 commented:
Outside of the recovery issue mentioned above, I've been doing a bunch of testing, and this has been looking pretty good, so I think it'll be close to ready once that is resolved.

dnldd commented:
Tested on the invalidation case on mainnet.

```sh
❯ dcrd --rpcuser=user --rpcpass=pass --profile=127.0.0.1:2555 --appdata=/home/dnldd/mh/node --datadi
r=/home/dnldd/mh/node/data --debuglevel=info --txindex --addrindex                                  
2021-03-23 12:16:21.476 [INF] DCRD: Version 1.7.0-pre (Go version go1.15.6 linux/amd64)             
2021-03-23 12:16:21.476 [INF] DCRD: Home dir: /home/dnldd/mh/node                                   
2021-03-23 12:16:21.476 [INF] DCRD: Loading block database from '/home/dnldd/mh/node/data/mainnet/bl
ocks_ffldb'                                                                                         
2021-03-23 12:16:21.476 [INF] DCRD: Creating profiling server listening on 127.0.0.1:2555           
2021-03-23 12:16:21.496 [INF] DCRD: Block database loaded                                           
2021-03-23 12:16:21.532 [INF] CHAN: Loading block index...                                          
2021-03-23 12:16:25.891 [INF] STKE: Stake database version 1 loaded                                 
2021-03-23 12:16:25.891 [INF] CHAN: UTXO cache initializing (max size: 150 MiB)...                  
2021-03-23 12:16:25.891 [INF] CHAN: UTXO cache initialization completed                             
2021-03-23 12:16:25.891 [INF] CHAN: Blockchain database version info: chain: 9, compression: 1, bloc
k index: 3                                                                                          
2021-03-23 12:16:25.891 [INF] CHAN: Best known header: height 4399, hash 00000000000022cedd1a1718817
df27adb896a6d71c95ac2f038369858e9280b                                                               
2021-03-23 12:16:25.891 [INF] CHAN: Chain state: height 4399, hash 00000000000022cedd1a1718817df27ad
b896a6d71c95ac2f038369858e9280b, total transactions 53409, work 3649431975704017200, progress 100.00
%                                                                                                   
2021-03-23 12:16:25.901 [INF] INDX: Transaction index is enabled                                    
2021-03-23 12:16:25.901 [INF] INDX: transaction index: recovering from tip 4499 (0000000000001cb5ca2
7d2141b32219c3822028e58ece373b8f47e018c100638)                                                      
2021-03-23 12:16:25.964 [INF] INDX: transaction index: index recovered to tip 4399 (00000000000022ce
dd1a1718817df27adb896a6d71c95ac2f038369858e9280b)                                                   
2021-03-23 12:16:25.964 [INF] INDX: Address index is enabled                                        
2021-03-23 12:16:25.964 [INF] INDX: address index: recovering from tip 4499 (0000000000001cb5ca27d21
41b32219c3822028e58ece373b8f47e018c100638)                                                          
2021-03-23 12:16:26.072 [INF] INDX: address index: index recovered to tip 4399 (00000000000022cedd1a
1718817df27adb896a6d71c95ac2f038369858e9280b)                                                       
2021-03-23 12:16:26.072 [INF] INDX: Exists address index is enabled                                 
2021-03-23 12:16:26.081 [INF] RPCS: RPC server listening on [::1]:9109                              
2021-03-23 12:16:26.081 [INF] RPCS: RPC server listening on 127.0.0.1:9109                          
2021-03-23 12:16:26.081 [INF] CMGR: Server listening on [::]:9108                                   
2021-03-23 12:16:26.081 [INF] CMGR: Server listening on 0.0.0.0:9108
```

rstaudt2 commented on blockchain/chain.go:
```diff
@@ -1067,6 +1246,10 @@ func (b *BlockChain) reorganizeChainInternal(target *blockNode) error {
 
+		// Ensure the associated spend journal of the disconnected block is not
+		// removed since it might be needed in a recovery process.
+		b.pruneFilter.AddEntry(block.Hash())
```
I think it would be helpful to add more of a description around who should be calling `AddEntry`/`RemoveEntry`, and when.

Since `AddEntry` increments the number of references to a block hash that we don't want to remove, shouldn't it be called by (or incremented based on) whoever still needs it (the indexers in this case).  It's being called every time here, which probably doesn't make sense if indexes aren't enabled (we would never remove the spend journal then).  Also, if another consumer of the spend journal used this, we'd need to increment it twice (for each consumer) rather than just having blockchain increment it once in the general case.

dnldd replied:
`AddEntry` is being called for `reorganizeChainInternal` because it's possible blocks initially disconnected the reorg process get reconnected. It's ideal to keep the spend journal data around for such a possibility. It is also possible to have out-of-sync disabled indexers that will need the spend journal entries of the disconnected blocks to recover up when enabled. It is also possible the reorganization was initiated by a call to `invalidateblock`, the spend journals for the disconnected blocks need to be kept around on the possibility the invalidated block is validated again via `reconsiderblock`. `reorganizeChainInternal` is the intend caller for `AddEntry` based on the reasons given above.

Regarding not immediately pruning the spend journal of indexes are not enabled, the cases not involving indexes mentioned above still apply. Also it is possible to have indexes disabled and out of sync. Once they are enabled they will still need the spend journal data of the disconnected blocks to recover back to the main chain. I will refine the check to exclude cases where the indexes do not exist yet but the other cases outlined earlier would still apply. 

I will update the comments for `AddEntry/RemoveEntry` as requested and take another look at improvements. I reckon possible events needing the spend journal data can collectively count as a single `AddEntry` call whereas confirmed out of sync indexes would be another another `AddEntry` call during a reorg.

rstaudt2 replied:
> AddEntry is being called for reorganizeChainInternal because it's possible blocks initially disconnected the reorg process get reconnected. It's ideal to keep the spend journal data around for such a possibility.

> It is also possible the reorganization was initiated by a call to invalidateblock, the spend journals for the disconnected blocks need to be kept around on the possibility the invalidated block is validated again via reconsiderblock. reorganizeChainInternal is the intend caller for AddEntry based on the reasons given above.

This is not accurate given that the spend journal is currently always removed when the block is disconnected.  The spend journal isn't needed to reconnect a block. 

My understanding of the problem that is trying to be solved here is:

- We are adding an asynchronous consumer of block connect/disconnect notifications (possibly multiple consumers in the future)
- The consumer needs the spend journal entry for disconnected blocks, but currently, that is removed immediately when disconnecting a block and is therefore not available

Given that, I think the goal of an ideal solution for this would be:

- The spend journal is retained **only** while there are still consumers that need it

In order to accomplish this, I think we would need to know whether each consumer has seen (or been notified) of the block being disconnected.  To build on your current solution, I think we could do something like:

- When a block is disconnected, "ask" each consumer if they still need the block
- If the consumer needs the block (i.e., it saw the block and hasn't disconnected it yet), increment the count
  - This should address the various cases outlined such as
    - Indexes never enabled (the disconnected block isn't needed)
    - Indexes enabled, then disabled (the disconnected block is needed only if the indexers connected that block previously)
- When that count goes to zero, remove the spend journal
- One caveat is this would require the consumer to track block hashes (which currently isn't an issue since the tx index tracks it and the addr index depends on the tx index)

The other thing that needs to be considered is how shutdowns are handled.  I suspect we may need to be tracking this in the database.

davecgh commented on signal.go:
```diff
@@ -20,9 +20,10 @@ var shutdownRequestChannel = make(chan struct{})
+// requests from shutdownRequestChannel.  It returns a context and its
+// cancellation function. The context is cancelled when either signal
+// is received.
+func shutdownListener() (context.Context, context.CancelFunc) {
```
This change needs to be reverted.  It is unused and not a safe way to about it anyway since its use would be result in the signal goroutine never terminating.

davecgh commented on server.go:
```diff
@@ -3289,7 +3290,7 @@ func setupRPCListeners() ([]net.Listener, error) {
 // decred network type specified by chainParams.  Use start to begin accepting
 // connections from peers.
-func newServer(ctx context.Context, listenAddrs []string, db database.DB, chainParams *chaincfg.Params, dataDir string) (*server, error) {
+func newServer(ctx context.Context, cancel context.CancelFunc, listenAddrs []string, db database.DB, chainParams *chaincfg.Params, dataDir string) (*server, error) {
```
As mentioned elsewhere, this change to take the cancel func needs to be reverted.

davecgh commented on dcrd.go:
```diff
@@ -61,7 +61,7 @@ func dcrdMain() error {
 	// triggered either from an OS signal such as SIGINT (Ctrl+C) or from
 	// another subsystem such as the RPC server.
-	ctx := shutdownListener()
+	ctx, cancel := shutdownListener()
```
Revert

davecgh commented on dcrd.go:
```diff
@@ -186,7 +186,7 @@ func dcrdMain() error {
 	// Create server.
 	lifetimeNotifier.notifyStartupEvent(lifetimeEventP2PServer)
-	svr, err := newServer(ctx, cfg.Listeners, db, cfg.params.Params,
+	svr, err := newServer(ctx, cancel, cfg.Listeners, db, cfg.params.Params,
```
Revert

davecgh commented on signal.go:
```diff
@@ -54,7 +55,7 @@ func shutdownListener() context.Context {
 	}()
 
-	return ctx
+	return ctx, cancel
```
Revert

davecgh reviewed (commented):
Several of the commit messages are not following the code contribution guidelines with respect to wrapping.

Also, please merge the review updates into the appropriate commits so effort is not wasted reviewing commits that are stale.

davecgh commented on cmd/addblock/addblock.go:
```diff
@@ -94,7 +95,8 @@ func realMain() error {
 	// The done channel returned from start will contain an error if
 	// anything went wrong.
-	importer, err := newBlockImporter(db, fi)
+	ctx, cancel := context.WithCancel(context.TODO())
```
This should be `context.Background()`.

davecgh commented on blockchain/chain.go:
```diff
@@ -310,6 +314,133 @@ type VoteInfo struct {
+//
+// This implements the indexers.PrevScripter interface.
+type prevScriptsSnapshot struct {
+	entries map[wire.OutPoint]*prevScript
```
Following up since there was an offline question about this and I think it'll be good to show why for other reviewers and contributors as well, this shows what I mean:

```Go
package main

import (
	"fmt"
	"runtime"
)

type prevScript struct {
	scriptVersion uint16
	pkScript      []byte
}

func main() {
	var memstats runtime.MemStats
	runtime.ReadMemStats(&memstats)
	mallocs := -memstats.Mallocs
	m := map[string]prevScript{
		"a": prevScript{},
		"b": prevScript{},
		"c": prevScript{},
		"d": prevScript{},
		"e": prevScript{},
		"f": prevScript{},
	}
	runtime.ReadMemStats(&memstats)
	mallocs += memstats.Mallocs
	fmt.Printf("mallocs without pointer: %d\n", mallocs)
	_ = m

	runtime.ReadMemStats(&memstats)
	mallocs = -memstats.Mallocs
	m2 := map[string]*prevScript{
		"a": &prevScript{},
		"b": &prevScript{},
		"c": &prevScript{},
		"d": &prevScript{},
		"e": &prevScript{},
		"f": &prevScript{},
	}
	runtime.ReadMemStats(&memstats)
	mallocs += memstats.Mallocs
	fmt.Printf("mallocs with pointer: %d\n", mallocs)
	_ = m2
}
```

Results:

```
mallocs without pointer: 0
mallocs with pointer: 6
```

dnldd commented on blockchain/indexers/addrindex.go:
```diff
@@ -1069,17 +1146,33 @@ func (idx *AddrIndex) UnconfirmedTxnsForAddress(addr dcrutil.Address) []*dcrutil
+
+	// The address index is an optional index. It depends on the
+	// transaction index and as a result synchronously updates with it.
+	sub, err := subscriber.Subscribe(idx, txIndexName)
```
`Subscribe` is taking the index as a parameter, it's being created first as a result.

dnldd commented:
Additional catchup tests on mainnet.

```sh
~/mh/node 11m 13s
❯ dcrd --rpcuser=user --rpcpass=pass --profile=127.0.0.1:2555 --appdata=/Users/dnldd/mh/node --datadir=/Users/dnldd/mh/node/data --debuglevel=info --txindex --dropaddrindex
2021-07-12 11:07:41.895 [INF] DCRD: Version 1.7.0-pre (Go version go1.16.5 darwin/arm64)
2021-07-12 11:07:41.895 [INF] DCRD: Home dir: /Users/dnldd/mh/node
2021-07-12 11:07:41.895 [INF] DCRD: Loading block database from '/Users/dnldd/mh/node/data/mainnet/blocks_ffldb'
2021-07-12 11:07:41.895 [INF] DCRD: Creating profiling server listening on 127.0.0.1:2555
2021-07-12 11:07:41.959 [INF] DCRD: Block database loaded
2021-07-12 11:07:41.959 [INF] CHAN: Loading UTXO database from '/Users/dnldd/mh/node/data/mainnet/utxodb'
2021-07-12 11:07:42.016 [INF] CHAN: UTXO database loaded
2021-07-12 11:07:42.017 [INF] INDX: Dropping all address index entries.  This might take a while...
2021-07-12 11:07:44.108 [INF] INDX: Deleted 1150770 keys (1150770 total) from address index
2021-07-12 11:07:47.675 [INF] INDX: Dropped address index
2021-07-12 11:07:47.675 [INF] DCRD: Gracefully shutting down the UTXO database...
2021-07-12 11:07:47.675 [INF] DCRD: Gracefully shutting down the block database...
2021-07-12 11:07:47.708 [INF] DCRD: Shutdown complete

~/mh/node 6s
❯ dcrd --rpcuser=user --rpcpass=pass --profile=127.0.0.1:2555 --appdata=/Users/dnldd/mh/node --datadir=/Users/dnldd/mh/node/data --debuglevel=info --txindex --addrindex    
2021-07-12 11:07:54.381 [INF] DCRD: Version 1.7.0-pre (Go version go1.16.5 darwin/arm64)
2021-07-12 11:07:54.381 [INF] DCRD: Home dir: /Users/dnldd/mh/node
2021-07-12 11:07:54.381 [INF] DCRD: Loading block database from '/Users/dnldd/mh/node/data/mainnet/blocks_ffldb'
2021-07-12 11:07:54.381 [INF] DCRD: Creating profiling server listening on 127.0.0.1:2555
2021-07-12 11:07:54.443 [INF] DCRD: Block database loaded
2021-07-12 11:07:54.443 [INF] CHAN: Loading UTXO database from '/Users/dnldd/mh/node/data/mainnet/utxodb'
2021-07-12 11:07:54.482 [INF] CHAN: UTXO database loaded
2021-07-12 11:07:54.546 [INF] CHAN: Loading block index...
2021-07-12 11:07:57.211 [INF] STKE: Stake database version 1 loaded
2021-07-12 11:07:57.211 [INF] CHAN: UTXO cache initializing (max size: 150 MiB)...
2021-07-12 11:07:57.212 [INF] CHAN: UTXO cache initialization completed
2021-07-12 11:07:57.212 [INF] CHAN: Blockchain database version info: chain: 10, compression: 1, block index: 3, spend journal: 3
2021-07-12 11:07:57.212 [INF] CHAN: UTXO database version info: version: 1, compression: 1, utxo set: 3
2021-07-12 11:07:57.212 [INF] CHAN: Best known header: height 571122, hash 0000000000000000429075ca28dab222128b7ed2457d89ac35b56d496cee7457
2021-07-12 11:07:57.212 [INF] CHAN: Chain state: height 69186, hash 0000000000000ea0b1ac52b8420db1686bdd6d8e24f1af4e78cfadcb60d15273, total transactions 1094906, work 185477045674204009936, progress 12.11%
2021-07-12 11:07:57.212 [INF] INDX: Transaction index is enabled
2021-07-12 11:07:57.213 [INF] INDX: Address index is enabled
2021-07-12 11:07:57.213 [INF] INDX: Exists address index is enabled
2021-07-12 11:07:57.213 [INF] INDX: Catching up from height 0 to 69186
2021-07-12 11:08:07.370 [INF] INDX: Indexed 25985 blocks in the last 10.15s (360148 transactions, height 25985, 2016-05-07 16:54:25 +0000 GMT)
2021-07-12 11:08:17.856 [INF] INDX: Indexed 22664 blocks in the last 10.42s (386835 transactions, height 48649, 2016-07-25 10:33:12 +0000 GMT)
2021-07-12 11:08:26.518 [INF] INDX: Caught up to height 69186
2021-07-12 11:08:26.526 [INF] CMGR: Server listening on [::]:9108
2021-07-12 11:08:26.526 [INF] RPCS: RPC server listening on [::1]:9109
2021-07-12 11:08:26.526 [INF] RPCS: RPC server listening on 127.0.0.1:9109
2021-07-12 11:08:26.526 [INF] CMGR: Server listening on 0.0.0.0:9108
2021-07-12 11:08:26.529 [INF] AMGR: Loaded 900 addresses from file '/Users/dnldd/mh/node/data/mainnet/peers.json'
2021-07-12 11:08:26.773 [INF] SYNC: New valid peer 51.38.68.167:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.1/)
2021-07-12 11:08:26.774 [INF] SYNC: Syncing headers to block height 571124 from peer 51.38.68.167:9108 (outbound)
2021-07-12 11:08:26.795 [INF] SYNC: New valid peer 104.248.33.247:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.0(rc3)/)
2021-07-12 11:08:26.901 [INF] SYNC: Processed 3 headers in the last 0.38s (progress 100.00%)
2021-07-12 11:08:26.901 [INF] SYNC: Initial headers sync complete (best header hash 00000000000000001a70abc6a809de892040dcb51dacf4ece890a56cce640de9, height 571124)
2021-07-12 11:08:26.901 [INF] SYNC: Syncing chain
2021-07-12 11:08:27.015 [INF] SYNC: New valid peer 154.30.192.70:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.2/)
2021-07-12 11:08:27.024 [INF] SYNC: New valid peer 159.65.98.117:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.5.1/)
2021-07-12 11:08:27.036 [INF] SYNC: New valid peer 159.65.98.117:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.5.1/)
2021-07-12 11:08:27.185 [INF] CMGR: 16 addresses found from seeder mainnet-seed.dcrdata.org
2021-07-12 11:08:27.195 [INF] CMGR: 16 addresses found from seeder mainnet-seed.jholdstock.uk
2021-07-12 11:08:27.209 [INF] SYNC: New valid peer 47.91.21.250:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.2/)
2021-07-12 11:08:27.335 [INF] CMGR: 16 addresses found from seeder mainnet-seed.planetdecred.org
2021-07-12 11:08:27.336 [INF] SYNC: New valid peer 176.117.195.164:9108 (outbound) (/dcrwire:0.3.0/dcrd:1.4.0/)
2021-07-12 11:08:27.347 [INF] CMGR: 16 addresses found from seeder mainnet-seed-1.decred.org
2021-07-12 11:08:27.683 [INF] SYNC: New valid peer 18.230.115.229:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.2/)
2021-07-12 11:08:27.951 [INF] CMGR: 16 addresses found from seeder mainnet-seed-2.decred.org
2021-07-12 11:08:36.952 [INF] SYNC: Processed 638 blocks in the last 10.05s (3036 transactions, 2558 tickets, 3115 votes, 71 revocations, height 69824, progress 12.23%)
^C2021-07-12 11:08:42.947 [INF] DCRD: Received signal (interrupt).  Shutting down...
2021-07-12 11:08:42.947 [WRN] SRVR: Server shutting down
2021-07-12 11:08:42.947 [INF] INDX: Index subscriber shutting down
2021-07-12 11:08:42.947 [WRN] RPCS: RPC server shutting down
2021-07-12 11:08:42.948 [INF] AMGR: Address manager shutting down
2021-07-12 11:08:42.948 [INF] RPCS: RPC server shutdown complete
2021-07-12 11:08:43.261 [INF] SRVR: Server shutdown complete
2021-07-12 11:08:43.261 [INF] DCRD: Gracefully shutting down the UTXO database...
2021-07-12 11:08:43.262 [INF] DCRD: Gracefully shutting down the block database...
2021-07-12 11:08:43.280 [INF] DCRD: Shutdown complete
```

dnldd commented:
Will squash the resolve commits once all issues are confirmed resolved.

davecgh reviewed (approved):
I've reviewed all updates and tested several cases such as dropping the transaction index and ensuring it also drops the dependent address index, performing a full reindex, and ensuring the blocking behavior works as intended.

The final review commit needs to be squashed into the appropriate commits to which the changes apply and then I'll get this merged.

davecgh commented:
Well, actually I see @rstaudt2 still has some changes requested that probably have been addressed, but I'd like to give him an opportunity to confirm before merging.

rstaudt2 reviewed (commented):
It looks like this is missing calling `RemoveSpendConsumerDependency` from the indexers (so spend journal entries are never removed for disconnected blocks when indexes are enabled).

Outside of that, this looks good overall to me.  I also confirmed that the panic that can be reproduced on `master` [with these steps](https://github.com/decred/dcrd/pull/2641#pullrequestreview-690679575) is now fixed.

davecgh commented:
I was doing some profiling on this with a full index and it appears it is holding onto memory that it shouldn't be.  See:

![image](https://user-images.githubusercontent.com/2115102/133520034-afdb102e-55ba-4a61-a3aa-7918a633843f.png)

rstaudt2 commented:
> I was doing some profiling on this with a full index and it appears it is holding onto memory that it shouldn't be. See:
> 
> ![image](https://user-images.githubusercontent.com/2115102/133520034-afdb102e-55ba-4a61-a3aa-7918a633843f.png)

Probably also due to `RemoveSpendConsumerDependency` not being called since that is what removes the dependencies being tracked in memory.

davecgh commented:
The relevant source from the profile:
```
github.com/decred/dcrd/blockchain/v4/indexers.(*IndexSubscriber).CatchUp

src/github.com/decred/dcrd/blockchain/indexers/indexsubscriber.go

Total:           0   171.37MB (flat, cum) 27.24%
275            .          .           			} 
276            .          .           		} else { 
277            .          .           			parent = cachedParent 
278            .          .           		} 
279            .          .            
280            .   169.36MB           		child, err := queryer.BlockByHash(hash) 
281            .          .           		if err != nil { 
282            .          .           			return err 
283            .          .           		} 
284            .          .            
285            .          .           		// Construct and send the index notification. 
286            .          .           		var prevScripts PrevScripter 
287            .     2.01MB           		err = db.View(func(dbTx database.Tx) error { 
```

In other words it's holding onto ~170M worth of block data.  I haven't dug into it at all, but my first guess would be that the indexing is creating a reference to the scripts within the blocks and because they're interior pointers, that naturally means the blocks can't be garbage collected.

dnldd commented:
`CatchUp` should not be holding on too long to pkScript references with the latest commit.
